jmclient/jmclient/wallet_service.py

```diff
@@ -0,0 +1,362 @@
+#! /usr/bin/env python
+from __future__ import (absolute_import, division,
+                        print_function, unicode_literals)
+
+from twisted.internet import reactor
+from twisted.internet import task
+from twisted.application.service import Service
+from numbers import Integral
+from jmclient.configure import jm_single, get_log
+
+"""Wallet service
+
+The purpose of this independent service is to allow
+running applications to keep an up to date, asynchronous
+view of the current state of its wallet, deferring any
+polling mechanisms needed against the backend blockchain
+interface here.
+"""
+
+jlog = get_log()
+
+class WalletService(Service):
+
+    def __init__(self, wallet):
+        # The two principal member variables
+        # are the blockchaininterface instance,
+        # which is currently global in JM but
+        # could be more flexible in future, and
+        # the JM wallet object.
+        # TODO support multiple wallets here.
+        self.bci = jm_single().bc_interface
+        self.wallet = wallet
+        self.synced = False
+
+        # Dicts of registered callbacks, by type
+        # and then by txinfo, for events
+        # on transactions.
+        self.callbacks = {}
+        self.callbacks["all"] = []
+        self.callbacks["unconfirmed"] = {}
+        self.callbacks["confirmed"] = {}
+
+        self.restart_callback = None
+
+        # transactions we are actively monitoring,
+        # i.e. they are not new but we want to track:
+        self.active_txids = []
+
+    def startService(self):
+        """ Encapsulates start up actions.
+        Here wallet sync.
+        """
+        Service.startService(self)
```

Comment:
this would look much less confusing if it was a call to `super()`

Reply:
Done in https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/359/commits/7637865a7f987e40f80a22209a5efd40ba6f34b5